### PR TITLE
chore: Hook the new KMS in the framework and Context

### LINKED ISSUE #1150: Hook the new KMS in framework and context
part of #985

### PR DESCRIPTION
**Pull request by Baha-sk**
closes #1150

Signed-off-by: Baha Shaaban <baha.shaaban@securekey.com>

### CODE REVIEW

File: pkg/framework/aries/default.go [hunk 1]
```diff
@@ -122,11 +150,102 @@ func setAdditionalDefaultOpts(frameworkOpts *Aries) error {
...
+	_, err = file.Write([]byte(masterLockEnc.Ciphertext))
+	if err != nil {
+		return err
+	}
 
 	return nil
 }
 
+func getPassphraseAndSalt(opts *Aries, keySize int) (string, []byte, error) {
+	passphrase := "secretPwd"
```
Comment (llorllale) [inline pkg/framework/aries/default.go]: Shouldn't this input be required from the user?

File: pkg/framework/aries/default.go [hunk 2]
```diff
@@ -122,11 +150,102 @@ func setAdditionalDefaultOpts(frameworkOpts *Aries) error {
...
+	if opts.salt != "" {
+		if len(opts.salt) != keySize {
+			return "", nil, fmt.Errorf("invalid salt option provided")
+		}
+
+		copy(salt, opts.salt)
+		opts.salt = ""
+	} else {
+		// default -  no salt
+		salt = nil
```
Comment (llorllale) [inline pkg/framework/aries/default.go]: Why not generate a new random salt?

Comment (Baha-sk) [reply pkg/framework/aries/default.go]: salt is optional, if the user doesn't provide one, the framework shouldn't generate one.

salt and passphrase are not stored anywhere... they are the `key` to unlock the master key which can be stored either in a file or an env variable.. for the default behaviour, the framework will generate a master key and secure it in a file so that future New() calls can still read it (of course the user must pass in the same passphrase and salt options they provided the first time to unlock the master key)

Comment (Baha-sk) [reply pkg/framework/aries/default.go]: also notice that `passphrase` and `salt` options are cleared once used to avoid keep in them in memory
see lines 230 and 240 above

File: pkg/framework/aries/default.go (hunk 1 above)
Comment (Baha-sk) [reply pkg/framework/aries/default.go]: to ease the creation of the Aries framework and not impact existing code (namely BDD tests) that don't use the new KMS (and SecretLock), the framework will use this default passphrase in case the passphrase option (and KMS/SecretLock options) are not passed

Comment (Baha-sk) [reply pkg/framework/aries/default.go]: to clarify, this is supposed to be temporary until we flip legacyKMS into KMS, then either passphrase or SecretLock options may become mandatory .. it's to be reviewed in the future

File: pkg/framework/aries/framework.go [hunk 3]
```diff
@@ -208,7 +219,23 @@ func WithProtocols(protocolSvcCreator ...api.ProtocolSvcCreator) Option {
...
 func WithLegacyKMS(k api.KMSCreator) Option {
 	return func(opts *Aries) error {
-		opts.kmsCreator = k
+		opts.legacyKMSCreator = k
+		return nil
+	}
+}
+
+// WithPassphrase injects a passphrase used to secure the master key used in a SecretLock service
+func WithPassphrase(s string) Option {
```
Comment (troyronda) [inline pkg/framework/aries/framework.go]: I don't think this shouldn't be top level - you should be providing your SecretLock. There is a particular SecretLock that is constructed with a passphrase and secret.

File: pkg/framework/aries/default.go (hunk 1 above)
Comment (llorllale) [reply pkg/framework/aries/default.go]: Right, so what should be the default secretlock implementation?

File: pkg/framework/aries/framework.go (hunk 3 above)
Comment (Baha-sk) [reply pkg/framework/aries/framework.go]: function removed

File: pkg/framework/aries/default.go [hunk 4]
```diff
@@ -7,8 +7,13 @@ SPDX-License-Identifier: Apache-2.0
 package aries
 
 import (
+	"crypto/sha256"
 	"fmt"
+	"io"
 	"net/http"
+	"os"
+
+	"github.com/google/tink/go/subtle/random"
```
[the remark this answers is not in the document]
Comment (Baha-sk) [reply pkg/framework/aries/default.go]: import removed

File: pkg/framework/aries/default.go (hunk 2 above)
Comment (Baha-sk) [reply pkg/framework/aries/default.go]: salt and passphrase are removed from framework.go and default.go

File: pkg/framework/aries/default.go (hunk 1 above)
Comment (Baha-sk) [reply pkg/framework/aries/default.go]: there is no more passphrase being passed as an option.

the default behaviour (if no SecretLock option is passed to `New()`) is to create a SecretLock with a master key file generated and stored `unprotected`  in a file. The default service will use the master key as is (without encryption/decryption of master key) to wrap KMS keys.

For a more secure approach, one must create a SecretLock option with secret expansion to protect the master key prior to storing in a path. Examples are in unit tests and a README document will be pushed soon to explain how to do this.

File: docs/kms_secretlock.md [hunk 5]
```diff
@@ -0,0 +1,143 @@
+# Aries New Secure KMS, SecretLock and Crypto capabilities
```
Comment (troyronda) [inline docs/kms_secretlock.md]: Remove “Aries New Secure” from the title.

File: docs/kms_secretlock.md [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+
+The aries-framework-go project has a mandate to protect keys from being compromised during usage or storage of keys. 
+This goal can achieved by hiding raw key bytes from being exchanged between the user and the framework. Also, keys
+should be secured and locked by a master key that is knwon to the user and protected by a passphrase. In order
```
Comment (troyronda) [inline docs/kms_secretlock.md]: Paraphrase is only an example protection.

File: docs/kms_secretlock.md [hunk 7]
```diff
@@ -0,0 +1,143 @@
+# Aries New Secure KMS, SecretLock and Crypto capabilities
+
+The aries-framework-go project has a mandate to protect keys from being compromised during usage or storage of keys.
```
Comment (troyronda) [inline docs/kms_secretlock.md]: We should just talk about the goal rather than Aries-framework-go mandates.

File: cmd/aries-js-worker/package-lock.json [hunk 8]
```diff
@@ -446,6 +446,14 @@
       "integrity": "sha512-Wm6ukoaOGJi/73p/cl2GvLjTI5JM1k/O14isD73YML8StrH/7/lRFgmg8nICZgD3bZZvjwCGxtMOD3wWNAu8cg==",
       "dev": true
     },
+    "axios": {
```
Comment (troyronda) [inline cmd/aries-js-worker/package-lock.json]: why was this file change as part of this PR?

File: README.md [hunk 9]
```diff
@@ -15,7 +15,7 @@ We aim to provide Go implementations of:
 
 - Decentralized identity standards including [W3C decentralized identifiers](https://w3c.github.io/did-core/) (DIDs), [W3C DID resolution](https://w3c-ccg.github.io/did-resolution/), and [W3C verifiable credentials](https://w3c.github.io/vc-data-model/).
 - Decentralized data communication protocols anchored in DIDs: [DIDComm](https://github.com/hyperledger/aries-rfcs/blob/master/concepts/0005-didcomm).
-- A pluggable dependency framework, where implementors can customize primitives via Service Provider Interfaces (SPIs). We have a "batteries included" model where default primitives are included -- such as a key management system (KMS), crypto, data storage, digital hub integration, etc.
+- A pluggable dependency framework, where implementors can customize primitives via Service Provider Interfaces (SPIs). We have a "batteries included" model where default primitives are included -- such as a [key management system (KMS), secret lock](docs/kms_secretlock.md), crypto, data storage, digital hub integration, etc.
```
Comment (troyronda) [inline README.md]: let's just include secret lock as part of KMS instead of an explicit mention here. 

- it's a bit odd to have a link with two of the items.
- secret lock isn't really a normal term AFAIK.

File: docs/kms_secretlock.md [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+The aries-framework-go project has a mandate to protect keys from being compromised during usage or storage of keys. 
+This goal can achieved by hiding raw key bytes from being exchanged between the user and the framework. Also, keys
+should be secured and locked by a master key that is knwon to the user and protected by a passphrase. In order 
+to achieve this, the following criteria must be met:
+ - The KMS service should support providing a key reference (or keyset handle) whenever a key is required. It must not
+   return or accept raw contents of keys within the framework and with the user.
+ - A SecretLock service is needed to protect keys by encrypting them when being stored.
+ - The SecretLock service must work with a master key that can be provided by the user.
+ - The master key should also be encrypted to protect it from being compromised.
+ - It should be unlocked by a passphrase known to the user.
```
Comment (troyronda) [inline docs/kms_secretlock.md]: Again - passphrase is only one of the mechanisms.

File: docs/kms_secretlock.md [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+ - Finally the crypto services must only accept key references provided by the KMS to execute operations.
+ 
+There are several ways to prepare and use the new KMS and SecreatLock services. The below sections describe each path:
```
Comment (troyronda) [inline docs/kms_secretlock.md]: typo. SecreatLock.

File: pkg/framework/context/context.go [hunk 12]
```diff
@@ -247,10 +254,18 @@ func WithProtocolServices(services ...dispatcher.ProtocolService) ProviderOption
 	}
 }
 
-// WithLegacyKMS injects a legacy kms service into the context.
+// WithLegacyKMS injects a legacy legacyKMS service into the context.
```
Comment (troyronda) [inline pkg/framework/context/context.go]: a legacy legacyKMS?

File: pkg/framework/context/context.go [hunk 13]
```diff
@@ -79,8 +81,13 @@ func (p *Provider) Service(id string) (interface{}, error) {
 	return nil, api.ErrSvcNotFound
 }
 
-// LegacyKMS returns a kms service.
+// LegacyKMS returns a legacyKMS service.
 func (p *Provider) LegacyKMS() legacykms.KeyManager {
+	return p.legacyKMS
+}
+
+// KMS returns a KMS service
```
Comment (troyronda) [inline pkg/framework/context/context.go]: We could at least expand the acronym in these comments :).

File: pkg/framework/aries/default.go [hunk 14]
```diff
@@ -122,7 +147,64 @@ func setAdditionalDefaultOpts(frameworkOpts *Aries) error {
...
 	}
 
-	// TODO add SecretLock creation here.. #1150
+	return nil
+}
+
+func createDefSecretLock(opts *Aries) error {
+	keySize := sha256.Size
+
+	file, err := os.OpenFile(defMKPath, os.O_WRONLY|os.O_CREATE, 0600)
```
Comment (troyronda) [inline pkg/framework/aries/default.go]: we are hardcoding a file lookup as the default secret lock?
this will break WASM.

Comment (troyronda) [reply pkg/framework/aries/default.go]: Also, in combination with the other default - local DB storage, this default local file secret lock is arguably not adding much.

File: docs/kms_secretlock.md [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+
+## No SecretLock/KMS options
+
+For a user who doesn't have a master key created, they can create an Aries instance without SecretLock/KMS options
+which will create a new unsecured master key behind the scenes and save it to local disk. This key will be used to
```
Comment (troyronda) [inline docs/kms_secretlock.md]: Doesn't work for WASM.

File: cmd/aries-js-worker/package-lock.json (hunk 8 above)
Comment (Baha-sk) [reply cmd/aries-js-worker/package-lock.json]: not sure, but it resulted from `make all`

File: pkg/framework/aries/default.go (hunk 14 above)
Comment (Baha-sk) [reply pkg/framework/aries/default.go]: another option would be to set an `env` variable, but it will be erased when the Go program shuts down. 

This means a new master key is generated on each aries.New() call without KMS/SecretLock options which makes keys stored in a local DB storage unlockable.

File: docs/kms_secretlock.md (hunk 5 above)
Comment (Baha-sk) [reply docs/kms_secretlock.md]: done

File: docs/kms_secretlock.md (hunk 6 above)
Comment (Baha-sk) [reply docs/kms_secretlock.md]: updated text to mention it as an example

File: docs/kms_secretlock.md (hunk 7 above)
Comment (Baha-sk) [reply docs/kms_secretlock.md]: updated text

File: README.md (hunk 9 above)
Comment (Baha-sk) [reply README.md]: removed secret lock from the link

File: docs/kms_secretlock.md (hunk 10 above)
Comment (Baha-sk) [reply docs/kms_secretlock.md]: updated to reflect this as an example

File: docs/kms_secretlock.md (hunk 11 above)
Comment (Baha-sk) [reply docs/kms_secretlock.md]: fixed

File: pkg/framework/aries/default.go (hunk 14 above)
Comment (Baha-sk) [reply pkg/framework/aries/default.go]: I will move this function to default_srv.go and use an the env reader master key in default_js_wasm.go

Comment (troyronda) [reply pkg/framework/aries/default.go]: @Baha-sk There are many options as SecretLock is an injectable interface. The problem is really that there might not be any acceptable default at the framework level (other than no-op).

There may be acceptable defaults for a particular target like cmd/agent-rest & the JS/WASM worker in cmd.

File: docs/kms_secretlock.md (hunk 15 above)
Comment (troyronda) [reply docs/kms_secretlock.md]: I think this section of text should be deferred to another PR once we figure out defaults for each target.

File: pkg/framework/aries/default_test.go [hunk 16]
```diff
@@ -20,6 +20,10 @@ func TestDefaultFramework(t *testing.T) {
 		defer cleanup()
 		dbPath = path
 
+		defMKPath = path + "/mk"
```
Comment (troyronda) [inline pkg/framework/aries/default_test.go]: I see that this is only tested during non-wasm builds.

Should be deferred into another PR.

File: pkg/framework/aries/default.go (hunk 14 above)
Comment (Baha-sk) [reply pkg/framework/aries/default.go]: sure, I'll add noop secretlock and make it as a framework default.

other specific targets may have their own secret lock instances (part of other PRs)..

I'll push another change in a bit...

File: pkg/framework/aries/default_test.go (hunk 16 above)
Comment (Baha-sk) [reply pkg/framework/aries/default_test.go]: removed

File: pkg/framework/aries/default.go (hunk 14 above)
Comment (Baha-sk) [reply pkg/framework/aries/default.go]: updated with noop SecretLock

File: pkg/framework/context/context.go (hunk 13 above)
Comment (Baha-sk) [reply pkg/framework/context/context.go]: done

File: pkg/framework/context/context.go (hunk 12 above)
Comment (Baha-sk) [reply pkg/framework/context/context.go]: updated

File: docs/kms_secretlock.md (hunk 15 above)
Comment (Baha-sk) [reply docs/kms_secretlock.md]: Updated with a more relevant comment.

Comment (Baha-sk) [reply docs/kms_secretlock.md]: created #1358 to add a default js-worker secretLock
and #1359 to add a default agent-rest secretLock
